### PR TITLE
Add sanity checks to all LCF data structure access

### PR DESCRIPTION
**Pull request by Ghabry**
to reduce crashes and exploit surface.

### CODE REVIEW

File: src/game_actor.cpp [hunk 1]
```diff
@@ -1133,18 +1213,36 @@ int Game_Actor::GetHitChance() const {
...
+	 * The following actor data is trusted after the initial verify:
+	 * - Class, SetClass filters invalid data
+	 * - Level is between 1 and 99
+	 * - Equipment
+	 * - Learned Skills
+	 * - ...
+	 *
+	 * Data::items (but not skills of items!)
+	 *
+	 */
```
Comment (Ghabry) [inline src/game_actor.cpp]: have to update this comment

File: src/game_actor.cpp [hunk 2]
```diff
@@ -459,9 +484,8 @@ int Game_Actor::CalculateExp(int level) const
 }
 
 void Game_Actor::MakeExpList() {
-	int final_level = Data::actors[actor_id - 1].final_level;
-	exp_list.resize(final_level, 0);;
-	for (int i = 1; i < final_level; ++i) {
+	exp_list.resize(99);
```
[the remark this answers is not in the document]
Comment (Ghabry) [reply src/game_actor.cpp]: Yes, this looks safe because GetLevel() is sanitized.

**Comment by Ghabry** [comment]
Fixed the reported problems and found more problems in the code.
Did a new commit to make the review easier.

File: src/window_battlestatus.cpp [hunk 3]
```diff
@@ -102,6 +103,7 @@ void Window_BattleStatus::RefreshGauge() {
 		}
 
 		for (int i = 0; i < item_max; ++i) {
+			// The always only contains valid battlers
```
Comment (carstene1ns) [inline src/window_battlestatus.cpp]: who?

**Comment by carstene1ns** [comment]
Looks good to me now apart from a few language issues in the comments, but this does not matter.
Not going to merge this, until at least a second review is available :grin:.

**Comment by Ghabry** [comment]
And that review should take at least 30 minutes ;)

File: src/scene_order.cpp [hunk 4]
```diff
@@ -19,6 +19,7 @@
 #include <algorithm>
 #include <string>
 #include <vector>
+#include <reader_util.h>
```
[the remark this answers is not in the document]
Comment (Ghabry) [reply src/scene_order.cpp]: Is really unused. 👍

File: src/background.cpp [hunk 5]
```diff
@@ -25,6 +25,8 @@
 #include "background.h"
 #include "bitmap.h"
 #include "main_data.h"
+#include "reader_util.h"
```
[the remark this answers is not in the document]
Comment (Ghabry) [reply src/background.cpp]: Fixed all your notes except this one because it already uses quotes

**Comment by Ghabry** [comment]
Don't merge this yet. Map data (Passages and Upper/Lower) is not verified...

**Comment by Ghabry** [comment]
Adjusted my sanity check code to the latest Battle CMD changes of 20kdc. Ready for merge now.

File: src/game_battler.cpp [hunk 6]
```diff
@@ -524,8 +562,11 @@ int Game_Battler::GetSpi() const {
 	int n = min(max(base_spi, 1), 999);
 
 	for (int16_t i : GetInflictedStates()) {
-		if(Data::states[i - 1].affect_spirit) {
-			n = AffectParameter(Data::states[i - 1].affect_type, base_spi);
+		// States are guaranteed to be valid
+		const RPG::State& state = *ReaderUtil::GetElement(Data::states, i);
+
```
Comment (carstene1ns) [inline src/game_battler.cpp]: Stray newline or missing in others.

Try to keep `GetAtk`, `GetDef()`, `GetSpi()` and `GetAgi()` the same.

File: src/game_interpreter.cpp [hunk 7]
```diff
@@ -2811,12 +2869,18 @@ bool Game_Interpreter::CommandReturnToTitleScreen(RPG::EventCommand const& /* co
...
 bool Game_Interpreter::CommandChangeClass(RPG::EventCommand const& com) { // code 1008
-	int class_id = com.parameters[2];
+	int class_id = com.parameters[2]; // 0: No class, 1+: Specific class
 	bool level1 = com.parameters[3] > 0;
 	int skill_mode = com.parameters[4]; // no change, replace, add
 	int stats_mode = com.parameters[5]; // no change, halve, level 1, current level
 	bool show = com.parameters[6] > 0;
 
+	const RPG::Class* cls = ReaderUtil::GetElement(Data::classes, class_id);
+	if (class_id > 0) {
```
Comment (carstene1ns) [inline src/game_interpreter.cpp]: I do not really understand this. What happens here?

File: src/scene_battle.cpp [hunk 8]
```diff
@@ -314,16 +342,42 @@ void Scene_Battle::AssignSkill(const RPG::Skill* skill) {
...
+					Output::Warning("AssignSkill: Item %d references invalid skill %d", skill_item->ID, skill_item->skill_id);
+					return;
+				}
+			}
+		}
+		default:
+			break;
+	}
+
+
```
Comment (carstene1ns) [inline src/scene_battle.cpp]: Stray newline.

**Comment by carstene1ns** [comment]
Additionally, I do not really like the additional braces you introduced in some `switch` `case`s.
IMO proper indentation is sufficient there.

We will need some code style document in the future to address such things.

**Comment by Ghabry** [comment]
Thx for the feedback. Do you mean braces like here:
https://github.com/Ghabry/easyrpg-player/blob/5fdc7d78e07f6c8196a6f3405cfbda25c7f3c7e9/src/game_interpreter.cpp#L2823-L2830

Unfortunately the braces are required here because you can't declare variables in a case without introducing a scope :/

File: src/game_actor.cpp [hunk 9]
```diff
@@ -421,17 +447,17 @@ int Game_Actor::GetBaseAgi() const {
 	return GetBaseAgi(true, true);
 }
 
-int Game_Actor::CalculateExp(int level) const
-{
+int Game_Actor::CalculateExp(int level) const {
+	const RPG::Class* klass = ReaderUtil::GetElement(Data::classes, GetData().class_id);
+
 	double base, inflation, correction;
 	if (GetData().class_id > 0) {
```
[the remark this answers is not in the document]
Comment (Ghabry) [reply src/game_actor.cpp]: Indeed. Looks like a bad rebase of the class_changed rename.

**Comment by carstene1ns** [comment]
Okay, did not know they were needed then.

**Comment by carstene1ns** [comment]
There seems to be a problem when returning to the gamebrowser from some games (maybe data is not cleared or cleared in the wrong order?): `Warning: Removing invalid item # from party`

**Comment by Ghabry** [comment]
Can you telll me a game where thus can be reproduced?

**Comment by fdelapena** [comment]
I guess it's the mentioned in EasyRPG/Player#1248

**Comment by carstene1ns** [comment]
Example games: Embric, Jojo, Lavender, Mondschein
To reproduce load a save or begin the game, press <kbd>F12</kbd> and choose Exit in the Menu.

**Comment by Ghabry** [comment]
Had to add special handling for level 0 everywhere to prevent nullptr-dereferences due to new sanity checks (guess before it read out of bounds and somehow survived).

The last commit was a random find while testing Ara Fell.

**Comment by Ghabry** [comment]
Sorry for the rebase, was to prevent a future conflict. Had to remove a sanity check in "SetupBattle" because the original code has a bug and can be completely deleted.  Will open an issue for this.

The last unreviewed commit was:
"Add exception in the sanity check and special handling for Level 0"

**Comment by carstene1ns** [comment]
Well, have further investigated the reason why the message pops up:
When returning from game to title scene via <kbd>F12</kbd>, `Data` is not cleared when [`Player::ResetGameObjects()`](https://github.com/Ghabry/easyrpg-player/blob/sanitycheck/src/scene_title.cpp#L58) is called. Thus, the items are still valid. However, when returning to the gamebrowser, they are still there, but [`Data::Clear()`](
https://github.com/Ghabry/easyrpg-player/blob/sanitycheck/src/scene_gamebrowser.cpp#L60-L61) has been called, so they are invalid.

As a fix, i think clearing the party items is necessary when leaving the game, either in `ResetGameObjects()` or where the party members are removed.

**Comment by Ghabry** [comment]
Looking forward to the bug reports. ;)

**Pull request merged**